core/src/main/scala/kafka/server/checkpoints/CheckpointFile.scala

```diff
@@ -56,13 +56,6 @@ class CheckpointFile[T](val file: File, version: Int, formatter: CheckpointFileF
 
         writer.flush()
         fileOutputStream.getFD().sync()
```

Comment:
Thanks for catching this. This becomes a problem due to the use of `LeaderEpochCache` in KIP-101. I have updated the patch to halt the system if `FileNotFoundException` is observed.

While it is possible catch the IOException thrown from this place and handle it, it will require a couple of tr/catch/handle distributed across the code. And non-trivial change is needed to avoid deadlock because `ReplicaFetcherThread.handleOffsetOutOfRange()` may also trigger this code.

I think the simplest approach is to just keep the existing code, i.e. halt the system if `FileNotFoundException` is thrown from here. I don't think it will affect the availability of JBOD deployment because according to Mayursh's comment, this `FileNotFoundException` is thrown if the broker is configured with RAID. We can re-investigate this problem if this is also an issue when broker uses JBOD.